<!doctype html> <html lang=en > <meta charset=UTF-8 > <meta name=viewport  content="width=device-width, initial-scale=1"> <link rel=stylesheet  href="/libs/highlight/style.css"> <link rel=stylesheet  href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.1/css/all.min.css" integrity="sha512-+4zCK9k+qNFUR5X+cKL9EIR+ZOhtIloNl9GIKS57V1MyNsYpYcUrUeQc9vNfzsWfV28IaLL3i96P9sdNyeRssA==" crossorigin=anonymous  /> <link rel=stylesheet  href="/css/franklin.css"> <link rel=icon  href="/assets/favicon.ico"> <title>How to turn DNA into Music</title> <header> <a class=title  href="/"><h2>0x0f0f0f - Alessandro Cheli</h2></a> <nav> <a href="/">Home</a> / <a href="/blog">Blog</a> / <a href="/tag/">Tags</a> <div class=social-media > <a style="color: #000000;" href="https://github.com/0x0f0f0f"><i class="fab fa-github"></i></a> <a style="color: #000000;" href="https://twitter.com/0x0f0f0f1"><i class="fab fa-twitter"></i></a> </div> </nav> </header> <main> <div class=container ><h1>How to turn DNA into Music</h1> <p>Published 04 April 2020</p> <p><b>Tags: </b><a href="/tag/functional-programming">#functional-programming</a>, <a href="/tag/ocaml">#ocaml</a>, <a href="/tag/minicaml">#minicaml</a>, <a href="/tag/bioinformatics">#bioinformatics</a>, <a href="/tag/dna">#dna</a></p><hr/> <h2 id=a_short_background ><a href="#a_short_background" class=header-anchor >A Short Background</a></h2> <p>I&#39;ve been playing with <a href="https://rosalind.info">bioinformatics algorithms</a> and DNA for a while during the COVID-19 lockdown. While reading a book chapter about machine translation, a weird idea came to my mind. I wondered if I could <em>procedurally translate DNA into music</em>. A few chapters later, the <a href="https://en.wikipedia.org/wiki/G&#37;C3&#37;B6del,_Escher,_Bach">author</a> also draws similarities between genes and music. It took me a few hours and a couple hundred lines of OCaml code to make a small program that translates <strong>FASTA</strong> files to playable MIDI files, mapped to an arbitrary music scale.</p> <blockquote> <p>In bioinformatics and biochemistry, the FASTA format is a text-based format for representing either nucleotide sequences or amino acid sequences, in which nucleotides or amino acids are represented using single-letter codes.</p> </blockquote> <p>The <code>dna2music</code> script reads a FASTA file from standard input and outputs an intermediate text file which contains sequences of notes, one by line, composed by their location, octave, the note played, velocity, duration and track.</p> <p>These text files can then converted to MIDI files by the handy <a href="https://github.com/alambicbicephale/txt2mid">txt2mid</a> tool, written in C&#43;&#43; by <a href="https://github.com/alambicbicephale">Jamil Alioui</a>.</p> <h2 id=but_how_does_it_sound_like ><a href="#but_how_does_it_sound_like" class=header-anchor >But How Does it Sound Like?</a></h2> <p>Figure: A screenshot of an Ardour session where the mp3 files were recorded. <img src="/assets/images/ardour.png" alt="Ardour Screenshot" /></p> <p>Just to be on the Coronavirus pandemic trend, here&#39;s how the SARS-Cov-2 genome sounds like, played in D# major at 120 beats per minute through the Open Source <a href="https://tytel.org/helm/">helm</a> software synthesizer.</p> <audio class=player  controls preload=none  style="width: 100%"> <source src="/assets/files/sarscov2.mp3" type="audio/mp3"> </audio> <p>Here&#39;s another take, this time on a small section of the genome of Cytomegalovirus, played in C.</p> <audio class=player  controls preload=none  style="width: 100%"> <source src="/assets/files/cytomegalovirus.mp3" type="audio/mp3"> </audio> <h2 id=how_does_the_conversion_work ><a href="#how_does_the_conversion_work" class=header-anchor >How does the conversion work?</a></h2> <p>The conversion is deterministic. This means that the same FASTA file will always produce the same MIDI file. I&#39;ve tried to partially replicate some processes that happen inside of living cells and then using modular arithmetics to extract meaningful interpretations from sequences of 1 or 4 aminoacids.</p> <p>At first, the DNA genome is read from standard input, expecting it is in the FASTA format. DNA genomes are written in an alphabet of 4 letters: A, C, G, T, which stand for adenine, cytosine, guanine and thymine. Those molecules are the nitrogen-containing biological compounds that form nucleosides, which in turn are components of nucleotides, the organic molecules, precisely monomers which form the DNA double helyx: a polymer.</p> <p>All the genome strings read from the FASTA file are merged into a big, single one:</p> <pre><code class=language-ocaml >let str &#61; String.concat &quot;&quot; @@ sndl @@ read_fasta stdin</code></pre>
<h2 id=the_first_semi-biological_step ><a href="#the_first_semi-biological_step" class=header-anchor >The First Semi-Biological Step</a></h2>
<p>Then, DNA is converted into RNA with a very simple function, which replaces the thymine symbol with the uracil symbol. In real life, this is done by an enzime called <a href="https://www.ncbi.nlm.nih.gov/books/NBK22085/">RNA polymerase</a>. RNAP locally opens the double-stranded DNA so that one strand of the exposed nucleotides can be used as a template for the synthesis of RNA in a process called transcription.</p>
<pre><code class=language-ocaml >let rna_of_dna s &#61; String.map &#40;fun c -&gt; match c with
  | &#39;T&#39; -&gt; &#39;U&#39;
  | a -&gt; a &#41; s</code></pre>
<p>The next step is converting the RNA into a chain of aminoacids. This is what is done inside of cells to produce proteins, the fundamental building blocks for almost everything inside all living organisms.</p>
<p>From a <a href="http://rosalind.info/problems/prot/">rosalind problem</a> :</p>
<blockquote>
<p>Just as nucleic acids are polymers of nucleotides, proteins are chains of smaller molecules called amino acids; 20 amino acids commonly appear in every species. Just as the primary structure of a nucleic acid is given by the order of its nucleotides, the primary structure of a protein is the order of its amino acids. Some proteins are composed of several subchains called polypeptides, while others are formed of a single polypeptide;</p>
</blockquote>
<p>Protein synthesis and folding are <strong>very complicated topics</strong>. Organelles inside of cells called ribosomes take chains of messenger RNA &#40;mRNA&#41; and a helping structure called transfer RNA, and examine the mRNA in pieces of 3 nucleotides &#40;symbols&#41; at a time, called <em>codons</em>. Since RNA is an &quot;<em>alphabet</em>&quot; of 4 symbols, there are 64 &#61; 4^3 possible strings of length 3. Since there are 20 aminoacids, multiple codons can encode the same aminoacid. The aminoacid corresponding to the currently examined codon of mRNA is the bond to the growing peptide chain.</p>
<p>Figure: The structure of the human haemoglobin protein. <img src="/assets/images/haemoglobin.png" alt="Human Haemoglobin" /></p>
<p>The same is done in <code>dna2music</code> to convert RNA to a sequence of aminoacids. There is a fundamental difference though: in biological cells, translation must start with the aminoacid methionine and ends when a stop codon is found &#40;encoded as the character &#39;0&#39; below&#41;. The stop codon corresponds to no aminoacid, but encodes a release factor: a protein that stops the translation and releases the peptide from the ribosome. This means that translation can start at any given part of the genome.</p>
<p>In <code>dna2music</code> instead, to produce a single track of music, the start and stop codons are ignored and the latter is treated as an additional imaginary aminoacid; the initial RNA string is trimmed to a length multiple of 3, to make sure that there are only valid codons. This biological heresy is to make sure that the resulting string is a single contiguous string of 21 aminoacids, 20 existing in reality and a &quot;fictionary&quot; one, corresponding to stop codons.</p>
<p>Here is the aminoacid conversion table:</p>
<pre><code class=language-ocaml >let encode_codon_protein codon &#61;
  if String.length codon &lt;&gt; 3 then
    failwith @@ codon ^ &quot; is not a valid codon&quot;
  else match codon with
  | &quot;AAA&quot; -&gt; &#39;K&#39; | &quot;AAC&quot; -&gt; &#39;N&#39; | &quot;AAG&quot; -&gt; &#39;K&#39; | &quot;AAU&quot; -&gt; &#39;N&#39;
  | &quot;ACA&quot; -&gt; &#39;T&#39; | &quot;ACC&quot; -&gt; &#39;T&#39; | &quot;ACG&quot; -&gt; &#39;T&#39; | &quot;ACU&quot; -&gt; &#39;T&#39;
  | &quot;AGA&quot; -&gt; &#39;R&#39; | &quot;AGC&quot; -&gt; &#39;S&#39; | &quot;AGG&quot; -&gt; &#39;R&#39; | &quot;AGU&quot; -&gt; &#39;S&#39;
  | &quot;AUA&quot; -&gt; &#39;I&#39; | &quot;AUC&quot; -&gt; &#39;I&#39; | &quot;AUG&quot; -&gt; &#39;M&#39; | &quot;AUU&quot; -&gt; &#39;I&#39;
  | &quot;CAA&quot; -&gt; &#39;Q&#39; | &quot;CAC&quot; -&gt; &#39;H&#39; | &quot;CAG&quot; -&gt; &#39;Q&#39; | &quot;CAU&quot; -&gt; &#39;H&#39;
  | &quot;CCA&quot; -&gt; &#39;P&#39; | &quot;CCC&quot; -&gt; &#39;P&#39; | &quot;CCG&quot; -&gt; &#39;P&#39; | &quot;CCU&quot; -&gt; &#39;P&#39;
  | &quot;CGA&quot; -&gt; &#39;R&#39; | &quot;CGC&quot; -&gt; &#39;R&#39; | &quot;CGG&quot; -&gt; &#39;R&#39; | &quot;CGU&quot; -&gt; &#39;R&#39;
  | &quot;CUA&quot; -&gt; &#39;L&#39; | &quot;CUC&quot; -&gt; &#39;L&#39; | &quot;CUG&quot; -&gt; &#39;L&#39; | &quot;CUU&quot; -&gt; &#39;L&#39;
  | &quot;GAA&quot; -&gt; &#39;E&#39; | &quot;GAC&quot; -&gt; &#39;D&#39; | &quot;GAG&quot; -&gt; &#39;E&#39; | &quot;GAU&quot; -&gt; &#39;D&#39;
  | &quot;GCA&quot; -&gt; &#39;A&#39; | &quot;GCC&quot; -&gt; &#39;A&#39; | &quot;GCG&quot; -&gt; &#39;A&#39; | &quot;GCU&quot; -&gt; &#39;A&#39;
  | &quot;GGA&quot; -&gt; &#39;G&#39; | &quot;GGC&quot; -&gt; &#39;G&#39; | &quot;GGG&quot; -&gt; &#39;G&#39; | &quot;GGU&quot; -&gt; &#39;G&#39;
  | &quot;GUA&quot; -&gt; &#39;V&#39; | &quot;GUC&quot; -&gt; &#39;V&#39; | &quot;GUG&quot; -&gt; &#39;V&#39; | &quot;GUU&quot; -&gt; &#39;V&#39;
  | &quot;UAC&quot; -&gt; &#39;Y&#39; | &quot;UAU&quot; -&gt; &#39;Y&#39; | &quot;UCA&quot; -&gt; &#39;S&#39; | &quot;UCC&quot; -&gt; &#39;S&#39;
  | &quot;UCG&quot; -&gt; &#39;S&#39; | &quot;UCU&quot; -&gt; &#39;S&#39; | &quot;UGC&quot; -&gt; &#39;C&#39; | &quot;UGG&quot; -&gt; &#39;W&#39;
  | &quot;UGU&quot; -&gt; &#39;C&#39; | &quot;UUA&quot; -&gt; &#39;L&#39; | &quot;UUC&quot; -&gt; &#39;F&#39; | &quot;UUG&quot; -&gt; &#39;L&#39;
  | &quot;UUU&quot; -&gt; &#39;F&#39; | &quot;UAG&quot; -&gt; &#39;0&#39; | &quot;UGA&quot; -&gt; &#39;0&#39; | &quot;UAA&quot; -&gt; &#39;0&#39;
  | _ -&gt; failwith &quot;invalid codon&quot; ;;</code></pre>
<h2 id=now_comes_the_music_making ><a href="#now_comes_the_music_making" class=header-anchor >Now Comes the Music Making&#33;</a></h2>
<p>This resulting aminoacid sequence is now a large string of an alphabet composed of 21 symbols, it&#39;s enough to turn it into music&#33; The symbol characters are converted to numbers from 0 to 20, so that the string is converted to a list of numbers. Modular arithmetics is used to restrict these numbers in smaller ranges. Notes in the chromatic scale go from 0 to 11, while notes in a major or minor scale go from 0 to 6. For example, notes in C Major scale, <code>C D E F G A
B</code>, can be mapped to numbers from 0 to 6. The same principle applies to all the major or minor scales.</p>
<p>The first aminoacid in the sequence corresponds to the number of notes that will be played in the current bar. It is equal to the aminoacid code number modulo 16, plus one. This means that it can range from 1 to 16. Let&#39;s call this number <code>n</code>.</p>
<p><code>n</code> following notes are read from the aminoacid sequence, if there are enough remaning aminoacids to compose them. After the notes are read, the conversion goes on to the next bar and reads another aminoacid that encodes again the number of notes <code>n</code> that will make the following bar. This mechanism makes sure that genomes of any length will always produce valid MIDI files that are not empty or cluttered with too many notes played altogether.</p>
<p>Notes are created by reading 4 aminoacids at a time. The first one encodes the location, or the delay in 16th notes from the start of the current bar. This is when the note will start playing, and is converted by multiplying 60 by the aminoacid code modulo 16: <code>&#40;num mod 16&#41; * 60</code>. In MIDI format a bar is 960 ticks, so a sixteenth note is 60 ticks long.</p>
<p>The second aminoacid in a note sequence encodes the duration of the note, which is read in the exact same way but excluding notes of length 0: <code>&#40;&#40;num mod 16&#41; &#43; 1&#41; * 60</code>. The third aminoacid in the quartet encodes the octave: <code>num mod 7</code>, while the fourth and last aminoacid making a note sequence finally encodes the note. This is done in a slightly more sophisticated way.</p>
<pre><code class=language-ocaml >if scale &gt;&#61; 0 then
    if minor then
        autoscale_minor scale &#40;note_in_scale_of_int note&#41;
    else
        autoscale_minor scale &#40;note_in_scale_of_int note&#41;
    else note_of_int note</code></pre>
<p>If a scale is passed as a command line argument to the program, the fourth aminoacid in the sequence encodes a note from 0 to 6, or <code>num mod 7</code>. The resulting note is then mapped automatically into a scale by the <code>autoscale_minor</code> or <code>autoscale_major</code> functions, which return only the chromatic note numbers from 0 to 11 in key with the scale passed as an argument. If automatical mapping to a scale is disabled, the aminoacid just encodes a note number from 0 to 11 in the chromatic scale. The latter case will obviously not sound musical at all&#33;</p>
<p>Here&#39;s a snippet of the text file resulting from SARS-Cov-2 genome-music conversion.</p>
<pre><code class=language-julia >N 1920 4 6 100 840 1
N 3120 5 8 100 60 1
N 3720 0 6 100 900 1
N 3960 2 6 100 600 1
N 5640 5 14 100 360 1
N 5340 6 14 100 480 1
N 4800 0 8 100 960 1
N 5040 2 16 100 900 1
N 5340 1 8 100 180 1
N 5640 4 8 100 960 1
N 5640 4 6 100 900 1
N 5700 4 9 100 60 1
N 4800 1 11 100 600 1
N 5160 0 13 100 840 1
N 5820 3 8 100 960 1
N 5760 2 8 100 960 1
N 5820 5 16 100 240 1
N 5760 7 6 100 360 1
N 5760 6 16 100 900 1
N 6660 6 8 100 120 1</code></pre>
<p>After printing out the whole sequence of notes, the output can be finally redirected into a text file and converted in format MIDI with the tool <code>txt2mid</code>, which reads exactly that format. This is how note lines are printed in <code>dna2music</code>:</p>
<h2 id=what_if_i_want_to_make_a_song_about_another_virus ><a href="#what_if_i_want_to_make_a_song_about_another_virus" class=header-anchor >What If I Want to Make a Song About Another Virus?</a></h2>
<p>If you want to use this buggy tool, you can download the source code from my <a href="https://github.com/0x0f0f0f/dna2music">GitHub profile</a>. You can also, just as an example, download the FASTA genome of Cytomegalovirus &#40;Human herpesvirus 5 strain AD169&#41; <a href="https://www.ncbi.nlm.nih.gov/nuccore/BK000394.5?report&#61;fasta">here</a>.</p>
<h3 id=building ><a href="#building" class=header-anchor >Building</a></h3>
<p>The project only has two dependencies: <code>g&#43;&#43;</code> and <code>ocaml</code> &gt;&#61; 4.03.0. You need to build <code>txt2mid</code> first:</p>
<pre><code class=language-bash >g&#43;&#43; txt2mid.cpp -o txt2mid</code></pre>
<h3 id=usage ><a href="#usage" class=header-anchor >Usage</a></h3>
<p>You can download genomes in FASTA format from the National Institutes of Health&#39;s <a href="https://www.ncbi.nlm.nih.gov/genbank/">GenBank</a> database, or from the <a href="https://genome.ucsc.edu/">UCSC Genome Browser</a>. After obtaining a FASTA file, you can pipe it into <code>dna2music</code> and redirect the output to a text file and then convert the text file to MIDI:</p>
<pre><code class=language-bash >cat /path/to/file.fasta | ocaml dna2music.ml &gt; file.txt
./txt2mid file.txt
# this will automatically create the MIDI file</code></pre>
<p><code>dna2music.ml</code> accepts two additional arguments regarding automatically mapping the notes to a certain key, to make the generated notes sound more musical altogether. If no additional argument is supplied key mapping will be skipped.</p>
<p>The first additional argument is the a string containing the key &#40;note&#41; in which the music will be composed, for example <code>C</code> or <code>E</code> are valid keys. To use sharp or flat key use the plus and minus symbols, for example <code>C&#43;</code> or <code>A-</code>.</p>
<p>The second additional argument is if the scale should be minor or not: <code>true</code> is minor and <code>false</code> is major. You can omit this argument to use the major scale. Other scales are not implemented. Have fun and stay at home during the lockdown&#33;</p>
</div>  </main>
  
    </div>  
    
    
        <script src="/libs/highlight/highlight.pack.js"></script>
<script>hljs.initHighlightingOnLoad();hljs.configure({tabReplace: '    '});</script>

    
    <footer>
      &copy; Alessandro Cheli.
      Made with <a href="https://github.com/tlienart/Franklin.jl">Franklin.jl</a>
      and <a href="https://julialang.org">The Julia Programming Language</a>.
    </footer>